File: cpp/questions/longestPalindrome.cpp
#include <bits/stdc++.h>

using namespace std;

class Solution {
private:
    std::pair<int, int> pairString(const std::string& s, const int& ss, int l, int r) {
        while (l>=1 && r<ss-1 && s[l-1]==s[r+1]) {
            --l;
            ++r;
        }
        return std::pair(l, r);
    }
public:
    string longestPalindrome(string s) {
        int ss = s.size();
        int as = 0;
        int al = 0;
        for (int i=0; i<ss; ++i) {
            auto [ns, ne] = pairString(s, ss, i, i);
            if (ne-ns > al) {
                as = ns;
                al = ne - ns;
            }
            if (i<ss-1 && s[i]==s[i+1]) {
                auto [ns, ne] = pairString(s, ss, i, i+1);
                if (ne-ns > al) {
                    as = ns;
                    al = ne - ns;
                }
            }
        }
        return s.substr(as, al+1);
    }
};

File: python/questions/longestPalindrome.py
class Solution:
    def longestPalindrome(self, s: str) -> str:
        ls = len(s)
        def ml(sp: int, ep: int) -> tuple[int, int]:
            while sp>=1 and ep<ls-1 and s[sp-1]==s[ep+1]:
                sp -= 1
                ep += 1
            return (sp, ep)
        ss = 0
        se = 0
        for i in range(ls):
            ns, ne = ml(i, i)
            if ne-ns > se-ss:
                ss = ns
                se = ne
            if i+1<ls and s[i]==s[i+1]:
                ns, ne = ml(i, i+1)
                if ne-ns > se-ss:
                    ss = ns
                    se = ne
        return s[ss:se+1]
